default/python/AI/ResourcesAI.py

```diff
@@ -148,115 +187,338 @@ def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: s
+                    continue
 
-        universe.updateMeterEstimates(unbaked_pids)
-        for pid, pinfo, planet in planet_infos:
-            po = pinfo.possible_output
-            industry_target = planet.currentMeterValue(fo.meterType.targetIndustry)
-            research_target = planet.currentMeterValue(fo.meterType.targetResearch)
-            influence_target = planet.currentMeterValue(fo.meterType.targetInfluence)
-            if planet.focus == INFLUENCE:
-                po[INFLUENCE] = (industry_target, research_target, influence_target)
-            else:
-                po[INFLUENCE] = (0, 0, 0)
-            if pinfo.planet.availableFoci and pinfo.current_focus != planet.focus:
-                fo.issueChangeFocusOrder(pid, pinfo.current_focus)  # put it back to what it was
+                # the increased population on the planet using this growth focus
+                # is mostly wasted, so ignore it for now.
+                pop = planet.currentMeterValue(fo.meterType.population)
+                pop_gain = potential_pop_increase - planet.habitableSize
+                if pop > pop_gain:
+                    _print_evaluation("would lose more pop (%.1f) than gain everywhere else (%.1f)." % (pop, pop_gain))
+                    continue
 
-        universe.updateMeterEstimates(unbaked_pids)
-        # Protection focus will give the same off-focus Industry and Research targets as Growth Focus
-        for pid, pinfo, planet in planet_infos:
-            po = pinfo.possible_output
-            po[PROTECTION] = po[GROWTH]
-            for focus in [INDUSTRY, RESEARCH, INFLUENCE, PROTECTION]:
-                debug(
-                    "Resource output of planet %s for focus %s is %f/%f/%f",
-                    pinfo.planet,
-                    focus,
-                    po[focus][0],
-                    po[focus][1],
-                    po[focus][2],
+                # If we have a computronium special here, then research focus will have higher priority.
+                if AIDependencies.COMPUTRONIUM_SPECIAL in planet.specials and RESEARCH in planet.availableFoci:
+                    _print_evaluation("has a usable %s" % AIDependencies.COMPUTRONIUM_SPECIAL)
+                    continue
+
+                _print_evaluation(
+                    "considered (pop %.1f, growth gain %.1f, current focus %s)" % (pop, pop_gain, pinfo.current_focus)
                 )
 
+                # add a bias to discourage switching out growth focus to avoid focus change penalties
+                if pinfo.current_focus == GROWTH:
+                    pop -= 4
+
+                ranked_planets.append((pop, pid, planet))
+
+            if not ranked_planets:
+                debug("  --> No suitable location found.")
+                continue
+
+            # sort possible locations by population in ascending order and set population
+            # bonus at the planet with lowest possible population loss.
```

Comment:
This sound like growth focuses always leads to population loss. 

I'd expect something like a planet with the most population increase by focus. 

Shouldn't we add population threshold for planets before sort?

Reply:
As I wrote in the PR, I only moved this function, I didn't write it. It is due to be changed as well, but that will be another PR.

Reply:
Could you point me to the places that need a review?